### PR TITLE
Add Observable#takeWhile() operator

### LINKED ISSUE #332: Observable.takeWhile operator
I tried my hands on creating a `takeWhile()` operator, but Kotlin Native doesn't seem to like it. My app was failing to compile with a strange error. Will it be possible for you to add one or help me send a PR by reviewing what's wrong with this code?

```kotlin
fun <T> Observable<T>.takeWhile(predicate: (T) -> Boolean): Observable<T> =
  observable { emitter ->
    val disposableWrapper = DisposableWrapper()
    emitter.setDisposable(disposableWrapper)

    subscribe(
        object : ObservableObserver<T>, CompletableCallbacks by emitter {
          override fun onSubscribe(disposable: Disposable) {
            disposableWrapper.set(disposable)
          }

          override fun onNext(value: T) {
            emitter.tryCatch(block = { predicate(value) }) {
              when {
                it -> emitter.onNext(value)
                else -> emitter.onComplete()
              }
            }
          }
        }
    )
  }
```

Error:

```
> Task :shared:linkDebugFrameworkIos
error: Invalid record (Producer: 'LLVM8.0.0svn' Reader: 'LLVM APPLE_1_1100.0.33.8_0')
1 error generated.
e: Compilation failed: The /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ command returned non-zero exit code: 1.
output:
 * Source files:
 * Compiler version info: Konan: 1.3.60 / Kotlin: 1.3.60
 * Output kind: FRAMEWORK
e: org.jetbrains.kotlin.konan.KonanExternalToolFailure: The /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ command returned non-zero exit code: 1.
```

More context here: https://kotlinlang.slack.com/archives/C3PQML5NU/p1576213492023600?thread_ts=1576134524.006800&cid=C3PQML5NU

### PR DESCRIPTION
**Pull request by saket**
Fixes #332. 

Warning: the k/n compiler was exploding somewhere in this operator in [my project](https://github.com/saket/press/pull/22). I've no idea why.

### CODE REVIEW

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/TakeWhilePredicateTest.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.badoo.reaktive.test.base.assertError
+import com.badoo.reaktive.test.observable.TestObservable
+import com.badoo.reaktive.test.observable.assertComplete
+import com.badoo.reaktive.test.observable.assertValues
+import com.badoo.reaktive.test.observable.onNext
+import com.badoo.reaktive.test.observable.test
+import kotlin.test.Test
+
+class TakeWhilePredicateTest : ObservableToObservableTests by ObservableToObservableTestsImpl({ takeWhile { true } }) {
```
Comment (saket) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/TakeWhilePredicateTest.kt]: I'm slightly surprised I only had to write 3 tests in this class. Is everything covered by generic base tests?

**Comment by arkivanov** [comment]
The PR title is `takeUntil` but you are adding `takeWhile` operator. Change the title?

**Comment by saket** [comment]
@arkivanov whoops, sorry for that. Done.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+fun <T> Observable<T>.takeWhile(predicate: (T) -> Boolean): Observable<T> =
+    observable { emitter ->
+        val disposableWrapper = DisposableWrapper()
+        emitter.setDisposable(disposableWrapper)
+
+        subscribe(
+            object : ObservableObserver<T>, CompletableCallbacks by emitter {
+                override fun onSubscribe(disposable: Disposable) {
+                    disposableWrapper.set(disposable)
```
Comment (arkivanov) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt]: No need for `DisposableWrapper` here, just call `emitter.set(disposable)`, check e.g. `map` operator. Would be nice to update `takeUntil` in the same way.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                }
+
+                override fun onNext(value: T) {
+                    emitter.tryCatch(block = { predicate(value) }) {
+                        when {
```
Comment (arkivanov) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt]: Please use `if-else` to keep consistency of the code.

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/TakeWhilePredicateTest.kt (hunk 1 above)
Comment (arkivanov) [reply reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/TakeWhilePredicateTest.kt]: That's ok, generic tests do the rest.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.tryCatch
+import com.badoo.reaktive.completable.CompletableCallbacks
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+
+/**
```
Comment (arkivanov) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt]: Let's just copy the text from [here](http://reactivex.io/documentation/operators/takewhile.html). And put the link as well: 
`See: [http://reactivex.io/documentation/operators/takewhile.html]`

Comment (saket) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt]: Done:

```kotlin
/**
 * Mirror items emitted by upstream until a specified condition becomes false.
 * See: [http://reactivex.io/documentation/operators/takewhile.html].
 */
```

The original doc was copied from RxJava, but this reads better.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt (hunk 3 above)
Comment (saket) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt]: Done

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt (hunk 2 above)
Comment (saket) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeWhilePredicate.kt]: Done

**Pull request merged**